File: src/com/nadmm/airports/wx/NoaaService.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

package com.nadmm.airports.wx;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.Serializable;
import java.net.URI;
import java.util.Date;
import java.util.zip.GZIPInputStream;

import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.utils.URIUtils;

import android.app.IntentService;
import android.content.Intent;

import com.nadmm.airports.utils.NetworkUtils;
import com.nadmm.airports.utils.SystemUtils;

public abstract class NoaaService extends IntentService {

    protected final String NOAA_HOST = "weather.aero";
    protected final String RADAR_HOST = "radar.weather.gov";

    protected final String NOAA_DATASERVER_PATH = "/dataserver1_4/httpparam";

    public static final String STATION_ID = "STATION_ID";
    public static final String CACHE_ONLY = "CACHE_ONLY";
    public static final String FORCE_REFRESH = "FORCE_REFRESH";
    public static final String RADIUS_NM = "RADIUS_NM";
    public static final String LOCATION = "LOCATION";
    public static final String HOURS_BEFORE = "HOURS_BEFORE";
    public static final String COORDS_BOX = "COORDS_BOX";
    public static final String TYPE = "TYPE";
    public static final String IMAGE_CODE = "IMAGE_CODE";
    public static final String RESULT = "RESULT";

    public static final String TYPE_TEXT = "TYPE_TEXT";
    public static final String TYPE_IMAGE = "TYPE_IMAGE";

    public static final String ACTION_GET_METAR = "flightintel.intent.action.GET_METAR";
    public static final String ACTION_GET_TAF = "flightintel.intent.action.GET_TAF";
    public static final String ACTION_GET_PIREP = "flightintel.intent.action.GET_PIREP";
    public static final String ACTION_GET_AIRSIGMET = "flightintel.intent.action.GET_AIRSIGMET";
    public static final String ACTION_GET_RADAR = "flightintel.intent.action.GET_RADAR";
    public static final String ACTION_GET_PROGCHART = "flightintel.intent.action.GET_PROGCHART";
    public static final String ACTION_GET_WIND = "flightintel.intent.action.GET_WIND";

    protected File mDataDir;
    protected int mAge;
    private HttpClient mHttpClient;

    public NoaaService( String name, long age ) {
        super( name );
        mHttpClient = NetworkUtils.getHttpClient();
        mDataDir = SystemUtils.getExternalDir( "wx/"+name );

        if ( !mDataDir.exists() ) {
            mDataDir.mkdirs();
        }

        // Remove any old files from cache first
        cleanupCache( mDataDir, age );
    }

    private void cleanupCache( File dir, long maxAge ) {
        // Delete all files that are older
        Date now = new Date();
        File[] files = dir.listFiles();
        for ( File file : files ) {
            long age = now.getTime()-file.lastModified();
            if ( age > maxAge ) {
                file.delete();
            }
        }
    }

    protected boolean fetchFromNoaa( String query, File file, boolean compressed ) 
            throws Exception {
        return fetchFromNoaa( NOAA_DATASERVER_PATH, query, file, compressed );
    }

    protected boolean fetchFromNoaa( String path, String query, File file, boolean compressed ) 
            throws Exception {
        return fetch( NOAA_HOST, path, query, file, compressed );
    }

    protected boolean fetch( String host, String path, String query, File file, boolean compressed ) 
            throws Exception {
        URI uri = URIUtils.createURI( "http", host, 80, path, query, null );
        return fetch( uri, file, compressed );
    }

    protected boolean fetch( URI uri, File file, boolean compressed ) 
            throws Exception {
        if ( !NetworkUtils.isNetworkAvailable( this ) ) {
            return false;
        }

        HttpGet get = new HttpGet( uri );
        HttpResponse response = mHttpClient.execute( get );

        int status = response.getStatusLine().getStatusCode();
        if ( status != HttpStatus.SC_OK ) {
            throw new Exception( response.getStatusLine().getReasonPhrase() );
        }

        byte[] buffer = new byte[ 16*1024 ];
        int count;
        FileOutputStream out = new FileOutputStream( file );
        InputStream in = response.getEntity().getContent();
        if ( compressed ) {
            in = new GZIPInputStream( in );
        }
        while ( ( count = in.read( buffer, 0, buffer.length ) ) != -1 ) {
            out.write( buffer, 0, count );
        }
        in.close();
        out.close();
        return true;
    }

    protected File getDataFile( String name ) {
        return new File( mDataDir, name );
    }

    protected void sendResultIntent( String action, String stationId, Serializable result ) {
        Intent intent = makeResultIntent( action, TYPE_TEXT );
        intent.putExtra( STATION_ID, stationId );
        intent.putExtra( RESULT, result );
        sendBroadcast( intent );
    }

    protected void sendResultIntent( String action, String code, File result ) {
        Intent intent = makeResultIntent( action, TYPE_IMAGE );
        intent.putExtra( IMAGE_CODE, code );
        if ( result.exists() ) {
            intent.putExtra( RESULT, result.getAbsolutePath() );
        }
        sendBroadcast( intent );
    }

    private Intent makeResultIntent( String action, String type ) {
        Intent intent = new Intent();
        intent.setAction( action );
        intent.putExtra( TYPE, type );
        return intent;
    }

}

File: src/main/java/com/nadmm/airports/wx/NoaaService.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2012-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.wx

import android.app.Service
import android.content.Intent
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import com.nadmm.airports.utils.NetworkUtils.doHttpsGet
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import java.io.File
import java.util.zip.GZIPInputStream

abstract class NoaaService(protected val name: String, protected val maxAgeMillis: Long) : Service() {
    private val serviceJob = SupervisorJob() // Use SupervisorJob for better error handling in children
    // Coroutine scope tied to Dispatchers.IO for background work
    protected val serviceScope = CoroutineScope(serviceJob + Dispatchers.IO + CoroutineName(name))

    private var _wxCache: WxCache? = null
    protected val wxCache get() = _wxCache!!

    override fun onCreate() {
        super.onCreate()

        _wxCache = WxCache(this, name, maxAgeMillis)

        Log.d(TAG, "onCreate() called for $name service")
    }

    override fun onBind(intent: Intent): IBinder? {
        // We are not using a bound service, so return null
        return null
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy() called for $name service")
        // Cancel all coroutines when the service is destroyed
        serviceJob.cancel() // This will cancel all coroutines launched in serviceScope
    }

    protected fun fetchFromNoaa(path: String, query: String?, file: File, compressed: Boolean=false): Boolean {
        return fetch(AWC_HOST, path, query, file, compressed)
    }

    protected fun fetch(host: String, path: String, query: String?, file: File, compressed: Boolean=false): Boolean {
        return doHttpsGet(
            this, host, path, query, file, null, null,
            if (compressed) GZIPInputStream::class.java else null
        )
    }

    object Events {
        private val _events = MutableSharedFlow<Bundle>()
        val events = _events.asSharedFlow()

        suspend fun post(result: Bundle) {
            _events.emit(result)
        }
    }

    companion object {
        private val TAG: String = NoaaService::class.java.simpleName

        protected const val AWC_HOST: String = "aviationweather.gov"

        const val STATION_ID: String = "STATION_ID"
        const val STATION_IDS: String = "STATION_IDS"
        const val CACHE_ONLY: String = "CACHE_ONLY"
        const val FORCE_REFRESH: String = "FORCE_REFRESH"
        const val RADIUS_SM: String = "RADIUS_SM"
        const val LOCATION: String = "LOCATION"
        const val IMAGE_TYPE: String = "IMAGE_TYPE"
        const val IMAGE_CODE: String = "IMAGE_CODE"
        const val TEXT_TYPE: String = "TEXT_TYPE"
        const val TEXT_CODE: String = "TEXT_CODE"
        const val RESULT: String = "RESULT"

        const val TYPE: String = "TYPE"
        const val TYPE_TEXT: String = "TYPE_TEXT"
        const val TYPE_GRAPHIC: String = "TYPE_GRAPHIC"
        const val ACTION: String = "ACTION"
        const val ACTION_GET_METAR: String = "flightintel.intent.wx.action.GET_METAR"
        const val ACTION_CACHE_METAR: String = "flightintel.intent.wx.action.CACHE_METAR"
        const val ACTION_GET_TAF: String = "flightintel.intent.wx.action.GET_TAF"
        const val ACTION_GET_PIREP: String = "flightintel.intent.wx.action.GET_PIREP"
        const val ACTION_GET_AIRSIGMET: String = "flightintel.intent.wx.action.GET_AIRSIGMET"
        const val ACTION_GET_PROGCHART: String = "flightintel.intent.wx.action.GET_PROGCHART"
        const val ACTION_GET_ICING: String = "flightintel.intent.action.wx.GET_ICING"
        const val ACTION_GET_FA: String = "flightintel.intent.action.wx.GET_FA"
        const val ACTION_GET_FB: String = "flightintel.intent.action.wx.GET_FB"
        const val ACTION_GET_GFA: String = "flightintel.intent.action.wx.GET_GFA"
    }
}
